internal/resolution/datasource/auth.go

```diff
@@ -0,0 +1,285 @@
+package datasource
+
+import (
+	"bytes"
+	"context"
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"slices"
+	"strings"
+	"sync/atomic"
+)
+
+type AuthenticationMethod int
+
+const (
+	AuthBasic AuthenticationMethod = iota
+	AuthBearer
+	AuthDigest
+)
+
+type AuthenticationInfo struct {
+	SupportedMethods []AuthenticationMethod // In order of preference, only one method will be attempted.
+
+	// AlwaysAuth determines whether to always send auth headers.
+	// If false, the server must respond with a WWW-Authenticate header which will be checked for supported methods.
+	// Must be set to false to use Digest authentication.
+	AlwaysAuth bool
+
+	Username    string // Basic & Digest, plain text.
+	Password    string // Basic & Digest, plain text.
+	BearerToken string
+	BasicAuth   string        // Base64-encoded username:password. Overrides Username & Password fields for Basic.
+	CnonceFunc  func() string `json:"-"` // Function used to generate cnonce string for Digest. OK to leave unassigned - mostly for use in tests.
+
+	lastUsed atomic.Value // The last-used authentication method - used when AlwaysAuth is false to automatically send Basic auth.
+}
+
+func (auth *AuthenticationInfo) GetRequest(ctx context.Context, httpClient *http.Client, url string) (*http.Response, error) {
```

Comment:
nit: just `Get` to mirror the `net/http` naming conventions? `GetRequest` is also a little confusing because it's not clear if we're meant to be getting a request object, or making a get request.

Reply:
Renamed to `Get()`, though I find `authInfo.Get()` a bit confusingly named as well.
I thought of `authInfo.DoGet()`, but then if we were to add the `http.Do()` method I wouldn't want `authInfo.DoDo()`...

Maybe the struct should be renamed. Something like `authenticatedHTTP`?